### ISSUE TITLE
NUnit4002 shouldn't trigger for unknown types

### ISSUE THREAD

**Issue by MaceWindu** [nunit/nunit.analyzers#856]
In my case it is triggered for value with `dynamic` type, but I suspect object also affected

```cs
// Is.Default will not be able to infer default value properly
Assert.That(someDynamic, Is.EqualTo(default(byte)));

// same for object
Assert.That(someObject, Is.EqualTo(default(byte)));
```

**Comment by mikkelbu** [comment]
Hi @MaceWindu 

Thanks for the report - we will look into it. I guess there are at least a couple of directions:
* either ignore unknown types
* add more types that we special case in `predefinedType`
* don't raise the diagnostic on every `default(X)` as we do now - see below https://github.com/nunit/nunit.analyzers/blob/cf2040d6459c4da6406cc44ff48baaf822ac114c/src/nunit.analyzers/UseSpecificConstraint/UseSpecificConstraintAnalyzer.cs#L74-L77
* ...

Ps. Can you provide a full example of the `dynamic` case? In the tests below then only `Test_Object_IsDefault` fails with
```
  Assert.That(someObject, Is.Default)
  Expected: default
  But was:  0
```

```csharp
        [Test]
        public void Test_Dynamic_IsEqual()
        {
            dynamic someDynamic = byte.MinValue;
            Assert.That(someDynamic, Is.EqualTo(default(byte)));
        }

        [Test]
        public void Test_Object_IsEqual()
        {
            object someObject = byte.MinValue;
            Assert.That(someObject, Is.EqualTo(default(byte)));
        }

        [Test]
        public void Test_Dynamic_IsDefault()
        {
            dynamic someDynamic = byte.MinValue;
            Assert.That(someDynamic, Is.Default);
        }

        [Test]
        public void Test_Object_IsDefault()
        {
            object someObject = byte.MinValue;
            Assert.That(someObject, Is.Default);
        }
```

**Comment by MaceWindu** [comment]
That's actually interesting. For positive case it works, but not for negative:

```cs
[Test]
public void Test_Dynamic_IsDefault_Negative()
{
	dynamic someDynamic = default(bool);
	Assert.That(someDynamic, Is.Not.Default);
}

[Test]
public void Test_Dynamic_IsEqual_Negative()
{
	dynamic someDynamic = default(bool);
	Assert.That(someDynamic, Is.Not.EqualTo(default(byte)));
}
```

not that I used `bool` vs `byte`, because for numeric types assert will accept default value (0) of any numeric type. Not sure if it is correct behavior.

**Comment by manfred-brands** [comment]
@MaceWindu I don't see where `Is.Default` doesn't do the right thing with `dynamic`?

```csharp
        [TestCase(default(bool))]
        [TestCase(default(byte))]
        [TestCase(default(int))]
        public void Success_Dynamic(dynamic d)
        {
            Assert.That(d, Is.Default);
        }

        [TestCase(true)]
        [TestCase(byte.MaxValue)]
        [TestCase(int.MinValue)]
        public void Failure_Dynamic(dynamic d)
        {
            Assert.That(d, Is.Not.Default);
        }
```

Yes, it doesn't pass with `object`:

```csharp
        [TestCase(default(bool))]
        [TestCase(default(byte))]
        [TestCase(default(int))]
        public void Test_Object_IsDefault(object o)
        {
            Assert.That(o, Is.Default);
        }
```

However, it depends on how to interpret `Is.Default`.
I noticed there was no documentation on [docs.nunit.org](https://docs.nunit.org/articles/nunit/writing-tests/constraints/Constraints.html), but the question what `Is.Default` on `object` should do was asked in the [nunit issue where the requirement was defined](https://github.com/nunit/nunit/issues/3457#issuecomment-787484525)

> I'm wondering, what should be the behavior for boxed values?
```csharp
object a = 0;
Assert.That(a, Is.Default);
```
> 
> What is the expected behavior for that code, should it pass (since 0 is `default(int)`) or fail (since 0 is not `default(object)`).

With the [answer](https://github.com/nunit/nunit/issues/3457#issuecomment-787530002):

> I think it should fail too. The method being called is `Assert.That<object>`, and `Is.Default` for `object` means something different than `Is.Default` for `int`.
>
> Would expect `Is.Default` to operate according to `default(T)`, where `T` comes from `Assert.That<T>(T actual, ...)`.

This then leave what the NUnit.Analyzer should do.
I agree it should not raise _NUnit4002_ when:
* The actual type is _declared_ as `object` and
*  the type _X_ in `Is.EqualTo(default(X))` is not `object`.

I will create a PR.

**Comment by MaceWindu** [comment]
My problem with `Is.Default` is that it loose asserted type information compared to `Is.Equal(default(X))`. If asserted valye type change  it will go unnoticed. E.g. in my case I test object, generated in runtime and if code generation will mess with types I will not notice it.

In any case, I am going to rewrite those asserts, because as I mentioned before it was a bit of surprise to me that `Assert(T1Value, Is.Equal(default(T2)))` will succeed if T1 and T2 are both numeric but different types. Probably correct behavior, but not what i wanted to assert.
